sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,459 @@
+                return val
+
+            raise TypeError(f"Unhandled type {type(var)}")
+
+        @classmethod
+        def _get_python_dict(cls, text: str) -> dict[Any, Any]:
+            """
+            Produce a python list from a text string describing a D-Bus array.
+            """
+            text = text.strip()
+            res = {}
+            while len(text) > 0:
+                (key, value), pos = cls._get_python_dict_entry(text)
+                res[key] = value
+                text = text[pos:].strip()
+
+            return res
+
+        @classmethod
+        def _get_python_list(cls, text: str) -> list[Any]:
+            text = text.strip()
+            res = []
+            while len(text) > 0:
+                value, pos = cls._get_python_value(text)
+                res.append(value)
+                text = text[pos:].strip()
+
+            return res
+
+        @classmethod
+        def _get_python_dict_entry(cls, text: str) -> tuple:
+            """
+            Produce a python dictionary from a text string describing a D-Bus dictionary.
+            """
+            key, pos1 = cls._get_python_value(text)
+            value, pos2 = cls._get_python_value(text[pos1:])
+            return (key, value), pos1 + pos2
+
+        @classmethod
+        def _get_python_value(cls, text: str) -> tuple:
+            """
+            Given a text string resulting from the invocation of ``dbus-send``,
+            parse it and produce the equivalent structure in python types.
+            """
+            text = text.replace("\n", "")
+            res = re.match("^ *u?int[0-9]{2} +([0-9]+)", text)
+            if res is not None:
+                return int(res.group(1)), res.span()[1]
+
+            res = re.match('^ *(string|object +path) +"([^"]*)"', text)
+            if res is not None:
+                return res.group(2), res.span()[1]
+
+            res = re.match("^ *boolean +(([Tt][Rr][Uu][Ee])|([Ff][Aa][Ll][Ss][Ee]))", text)
+            if res is not None:
+                return (res.group(1).lower() == "true"), res.span()[1]
+
+            res = re.match("^ *variant +(.+) *$", text)
+            if res is not None:
+                value, _ = cls._get_python_value(res.group(1))
+                return Variant(value), res.span()[1]
+
+            res = re.match("^ *dict entry +\\((.+)\\)", text)
+            if res is not None:
+                return cls._get_python_dict_entry(res.group(1)), res.span()[1]
+
+            # dbus-send doesn't return "dict {...}" but "array [ dict entry (...) ... ]"
+            res = re.match("^ *array +\\[( *dict entry +.+)\\] *$", text)
+            if res is not None:
+                return cls._get_python_dict(res.group(1)), res.span()[1]
+
+            res = re.match("^ *array +\\[(.+)\\] *$", text)
+            if res is not None:
+                return cls._get_python_list(res.group(1)), res.span()[1]
+
+            raise TypeError(f"Unhandled D-Bus type {text}")
```

Comment:
Move these to standalone classes DBUSTypeArray, DBUSTypeDict, DBUSTypeScalar with parser/builder/getter/setter?

Reply:
Done. I also improved the parsing a bit.